File: Task2/task.cpp
#include <iostream>
#include <vector>
#include <set>
#include <cstdlib>
#include <ctime>

using namespace std;

const int MAXN = 1001;

vector<int> graph[MAXN];
bool visited[MAXN];

void dfs(int node) {
    visited[node] = true;
    for (int neighbor : graph[node]) {
        if (!visited[neighbor])
            dfs(neighbor);
    }
}

set<pair<int, int>> generateEdges(int N, int M) {
    set<pair<int, int>> edges;
    srand(time(0));

    while ((int)edges.size() < M) {
        int u = rand() % N + 1;
        int v = rand() % N + 1;
        if (u == v) continue;
        if (u > v) swap(u, v);

        edges.insert({u, v});
    }

    return edges;
}

int main() {
    int N, M;
    cout << "Введите количество узлов (N <= 1000): ";
    cin >> N;
    cout << "Введите количество связей (M <= 1000): ";
    cin >> M;

    if (N > 1000,  M > 1000,  N < 1 || M < 0) {
        cout << "Ошибка: N и M должны быть в пределах 1..1000" << endl;
        return 1;
    }

    auto edges = generateEdges(N, M);

    cout << "\nСгенерированные связи:\n";
    for (auto [u, v] : edges) {
        cout << u << " " << v << endl;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    int components = 0;
    for (int i = 1; i <= N; ++i) {
        if (!visited[i]) {
            dfs(i);
            components++;
        }
    }

    cout << "\nМинимальное число дополнительных связей: " << (components - 1) << endl;

    return 0;
}

File: Task2/task.py
import random

def dfs(graph, start_node, visited):
    """DFS для поиска компоненты связности"""
    stack = [start_node]
    
    while stack:
        node = stack.pop()
        if node not in visited:
            visited.add(node)
            # Добавляем непосещенных соседей в стек
            for neighbor in graph[node]:
                if neighbor not in visited:
                    stack.append(neighbor)

def count_connected_components(graph, n):
    """Подсчитывает количество компонент связности в графе"""
    visited = set()
    components = 0
    
    for node in range(1, n + 1):
        if node not in visited:
            dfs(graph, node, visited)
            components += 1
    
    return components

def generate_random_edges(n, m):
    """Генерирует m случайных рёбер для графа с n узлами"""
    edges = []
    used_edges = set()
    
    while len(edges) < m:
        u = random.randint(1, n)
        v = random.randint(1, n)
        
        # Избегаем петель и дубликатов
        if u != v and (u, v) not in used_edges and (v, u) not in used_edges:
            edges.append((u, v))
            used_edges.add((u, v))
    
    return edges

# Считываем входные данные
n = int(input("Введите количество узлов сети: "))
m = int(input("Введите число линий связи (ребер): "))

# Генерируем случайные рёбра
edges = generate_random_edges(n, m)

# Создаем список смежности для графа
graph = [[] for _ in range(n + 1)]  # Индексы от 1 до n

# Добавляем рёбра в граф
for u, v in edges:
    graph[u].append(v)
    graph[v].append(u)  # Граф ненаправленный

# Выводим сгенерированные узлы
print("Сгенерированные узлы:")
for u, v in edges:
    print(f"{u} {v}")

# Находим количество компонент связности
components = count_connected_components(graph, n)

# Выводим минимальное количество рёбер для объединения
print(f"Количество необходимых связей: {components - 1}")

